### PR TITLE
MULE-19620: Fixed ClassLoader Leak on applications using 

### PR DESCRIPTION
**Pull request by nicomz**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 1]
```diff
@@ -47,9 +46,9 @@
 
   @Test
   public void testIsConcrete() throws Exception {
-    assertTrue(ClassUtils.isConcrete(Orange.class));
-    assertTrue(!ClassUtils.isConcrete(Fruit.class));
-    assertTrue(!ClassUtils.isConcrete(AbstractFruit.class));
+    assertThat(ClassUtils.isConcrete(Orange.class),is(true));
+    assertThat(ClassUtils.isConcrete(Fruit.class),is(false));
+    assertThat(ClassUtils.isConcrete(AbstractFruit.class),is(false));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:51]: static imports

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 2]
```diff
@@ -62,9 +61,8 @@ public void testIsConcrete() throws Exception {
   @Test
   public void testLoadClass() throws Exception {
     Class clazz = ClassUtils.loadClass("java.lang.String", getClass());
-    assertNotNull(clazz);
-
-    assertEquals(clazz.getName(), "java.lang.String");
+    assertThat(clazz,is(notNullValue()));
+    assertThat(clazz.getName(),is( "java.lang.String"));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:65]: formate

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 3]
```diff
@@ -77,23 +75,23 @@ public void testLoadClass() throws Exception {
...
-    assertSame(ClassUtils.loadClass("long", getClass()), Long.TYPE);
-    assertSame(ClassUtils.loadClass("short", getClass()), Short.TYPE);
+    assertThat(ClassUtils.loadClass("boolean", getClass()),sameInstance(Boolean.TYPE));
+    assertThat(ClassUtils.loadClass("byte", getClass()), sameInstance(Byte.TYPE));
+    assertThat(ClassUtils.loadClass("char", getClass()), sameInstance(Character.TYPE));
+    assertThat(ClassUtils.loadClass("double", getClass()), sameInstance(Double.TYPE));
+    assertThat(ClassUtils.loadClass("float", getClass()), sameInstance(Float.TYPE));
+    assertThat(ClassUtils.loadClass("int", getClass()), sameInstance(Integer.TYPE));
+    assertThat(ClassUtils.loadClass("long", getClass()), sameInstance(Long.TYPE));
+    assertThat(ClassUtils.loadClass("short", getClass()), sameInstance(Short.TYPE));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:85]: static imports + format

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 4]
```diff
@@ -107,17 +105,17 @@ public void testLoadClassOfType() throws Exception {
   @Test
   public void testInstanciateClass() throws Exception {
     Object object = ClassUtils.instantiateClass("org.mule.tck.testmodels.fruit.Orange");
-    assertNotNull(object);
-    assertTrue(object instanceof Orange);
+    assertThat(object,is(notNullValue()));
+    assertThat(object instanceof Orange,is(true));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:109]: use the `instanceOf` matcher

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 5]
```diff
@@ -107,17 +105,17 @@ public void testLoadClassOfType() throws Exception {
...
 
     object = ClassUtils.instantiateClass("org.mule.tck.testmodels.fruit.FruitBowl", new Apple(), new Banana());
-    assertNotNull(object);
-    assertTrue(object instanceof FruitBowl);
+    assertThat(object,is(notNullValue()));
+    assertThat(object instanceof FruitBowl,is(true));;
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:113]: use the `instanceOf` matcher

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 6]
```diff
@@ -133,33 +131,34 @@ public void testGetParameterTypes() throws Exception {
     FruitBowl bowl = new FruitBowl();
 
     Class[] classes = ClassUtils.getParameterTypes(bowl, "apple");
-    assertNotNull(classes);
-    assertEquals(1, classes.length);
-    assertEquals(Apple.class, classes[0]);
+    assertThat(classes,is(notNullValue()));
+    assertThat(classes.length,is(equalTo(1)));
+    assertThat(classes[0],is(equalTo(Apple.class)));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:136]: use array matchers

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 7]
```diff
@@ -133,33 +131,34 @@ public void testGetParameterTypes() throws Exception {
...
+
 
     classes = ClassUtils.getParameterTypes(bowl, "invalid");
-    assertNotNull(classes);
-    assertEquals(0, classes.length);
+    assertThat(classes,is(notNullValue()));
+    assertThat(classes.length,is(equalTo(0)));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:141]: use an array legth matcher

File: core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java [hunk 8]
```diff
@@ -175,59 +174,88 @@ public void testEqual() {
...
     Object a = new HashBlob(1);
     Object b = new HashBlob(2);
-    assertTrue(ClassUtils.hash(new Object[] {a, b, a, b}) == ClassUtils.hash(new Object[] {a, b, a, b}));
-    assertFalse(ClassUtils.hash(new Object[] {a, b, a}) == ClassUtils.hash(new Object[] {a, b, a, b}));
-    assertFalse(ClassUtils.hash(new Object[] {a, b, a, a}) == ClassUtils.hash(new Object[] {a, b, a, b}));
-    assertFalse(ClassUtils.hash(new Object[] {b, a, b, a}) == ClassUtils.hash(new Object[] {a, b, a, b}));
+    assertThat(ClassUtils.hash(new Object[] {a, b, a, b}) == ClassUtils.hash(new Object[] {a, b, a, b}),is(true));
+    assertThat(ClassUtils.hash(new Object[] {a, b, a}) == ClassUtils.hash(new Object[] {a, b, a, b}),is(false));
+    assertThat(ClassUtils.hash(new Object[] {a, b, a, a}) == ClassUtils.hash(new Object[] {a, b, a, b}),is(false));
+    assertThat(ClassUtils.hash(new Object[] {b, a, b, a}) == ClassUtils.hash(new Object[] {a, b, a, b}),is(false));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/util/ClassUtilsTestCase.java:195]: use an equeals matcher

File: core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java [hunk 9]
```diff
@@ -372,36 +380,133 @@ public static Class loadClass(final String className, final ClassLoader classLoa
...
+   * @param <T>         the type of the field
+   * @return the value of the field
+   * @throws NoSuchFieldException   when the field is not found
+   * @throws IllegalAccessException when the field is not reachable.
+   */
+  public static <T> T getStaticFieldValue(Class<?> targetClass, String fieldName, boolean recursive)
+      throws NoSuchFieldException, IllegalAccessException {
+    Field field = getField(targetClass, fieldName, recursive);
+    boolean isAccessible = field.isAccessible();
+    if (!Modifier.isStatic(field.getModifiers())) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java:429]: static import

File: core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java [hunk 10]
```diff
@@ -372,36 +380,133 @@ public static Class loadClass(final String className, final ClassLoader classLoa
...
 
-        return;
-      } catch (NoSuchFieldException e) {
-        // ignore and look in superclass
-        if (recursive) {
-          clazz = clazz.getSuperclass();
-        } else {
-          break;
-        }
+    if (Modifier.isFinal(field.getModifiers())) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java:457]: static import

File: core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java [hunk 11]
```diff
@@ -372,36 +380,133 @@ public static Class loadClass(final String className, final ClassLoader classLoa
...
+   * @param recursive   flags to lookup the field in subclasses or not
+   * @throws NoSuchFieldException   when the field does not exists
+   * @throws IllegalAccessException when the field is not reachable
+   */
+  public static void setStaticFieldValue(Class<?> targetClass, String fieldName, Object value, boolean recursive)
+      throws NoSuchFieldException, IllegalAccessException {
+    Field field = getField(targetClass, fieldName, recursive);
+    boolean isAccessible = field.isAccessible();
+
+    if (Modifier.isFinal(field.getModifiers())) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/util/ClassUtils.java:493]: static import

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/IBMMQResourceReleaser.java [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+import static java.lang.management.ManagementFactory.getPlatformMBeanServer;
+import static org.mule.runtime.core.api.util.ClassUtils.getField;
+import static org.mule.runtime.core.api.util.ClassUtils.getStaticFieldValue;
+import static org.mule.runtime.core.api.util.ClassUtils.loadClass;
+import static org.mule.runtime.core.api.util.ClassUtils.setStaticFieldValue;
+
+/**
+ * A Utility class for releasing all the known references that may lead to a ClassLoader leak.
+ */
+public class IBMMQResourceReleaser implements ResourceReleaser {
```
Comment (elrodro83) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/IBMMQResourceReleaser.java:40]: add TODOs to extract common functionality with other releasers

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by nicomz** [comment]
--validate

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**